<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.324">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>A cycling, progenitor-like cell population at the base of atypical teratoid rhabdoid tumor subtype differentiation trajectories - CNV analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../analysis/snRNAseq/05_Supervised_Annotation.html" rel="next">
<link href="../../analysis/snRNAseq/03_Normalization_and_Integration.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script><style>html{ scroll-behavior: smooth; }</style>
<script type="text/javascript" src="./intense.js"></script>
<meta property="og:title" content="A cycling, progenitor-like cell population at the base of atypical teratoid rhabdoid tumor subtype differentiation trajectories - CNV analysis">
<meta property="og:site-name" content="A cycling, progenitor-like cell population at the base of atypical teratoid rhabdoid tumor subtype differentiation trajectories">
<meta name="twitter:title" content="A cycling, progenitor-like cell population at the base of atypical teratoid rhabdoid tumor subtype differentiation trajectories - CNV analysis">
<meta name="twitter:card" content="summary">
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../analysis/snRNAseq/01_Quality_Control.html">Analysis | snRNAseq</a></li><li class="breadcrumb-item"><a href="../../analysis/snRNAseq/04_CNV_analysis.html">CNV analysis</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../../">A cycling, progenitor-like cell population at the base of atypical teratoid rhabdoid tumor subtype differentiation trajectories</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/DrostGroup/ATRT_SC_Het" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false">
 <span class="menu-text">Abstract</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Co-First Authors</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Analysis | snRNAseq</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/snRNAseq/01_Quality_Control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quality Control</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/snRNAseq/02_Doublet_Removal.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Doublet Removal</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/snRNAseq/03_Normalization_and_Integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Normalization and Integration</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/snRNAseq/04_CNV_analysis.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">CNV analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/snRNAseq/05_Supervised_Annotation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Supervised Annotation</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false">
 <span class="menu-text">Analysis | snATACseq</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/snATACseq/01_Analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">snATACseq analysis</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">
 <span class="menu-text">Analysis | Xenium</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/Xenium/01_Analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Xenium analysis</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
 <span class="menu-text">Analysis | bulkRNAseq</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/bulkRNAseq/01_GenerateDESeq2Dataset.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Generate DESeq2 Dataset</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/bulkRNAseq/02_QC_Assessment.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quality Control</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/bulkRNAseq/03_DE_Analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">DE analysis</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false">
 <span class="menu-text">Figures</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure 1</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure 2</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure 3</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure 4</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S1</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S1</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S2</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S4</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S5</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S6.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S6</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S7.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S7</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S8.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S8</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S9.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S9</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S10.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S10</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S11.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S11</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S12.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S12</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">CNV analysis</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Methodology from: https://www.cell.com/cell-reports-medicine/pdf/S2666-3791(23)00426-3.pdf</span></span>
<span></span>
<span></span>
<span><span class="co"># Subset Seurat object to only contain tumor cells.</span></span>
<span></span>
<span><span class="va">sample</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="va">sample</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span>, <span class="st">"Assay"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate a new metadata column storing the mapping cell-metacell.</span></span>
<span><span class="va">sample</span><span class="op">[[</span><span class="st">"metacell_mapping"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"not_mapped"</span></span>
<span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu">Idents</span><span class="op">(</span><span class="va">sample</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sample</span><span class="op">$</span><span class="va">TME_annotation</span></span>
<span></span>
<span><span class="co"># Will store all the metacells.</span></span>
<span><span class="va">whole_metacells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span>, row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Will store the complete annotation for the metacells.</span></span>
<span><span class="va">whole_annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>cluster_names <span class="op">=</span> <span class="st">"test"</span>, row.names <span class="op">=</span> <span class="st">"test"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">meta_counter</span> <span class="op">&lt;-</span> <span class="fl">0</span> <span class="co"># To keep a count of the metacells that are created.</span></span>
<span><span class="va">metacell_content</span> <span class="op">&lt;-</span> <span class="fl">5</span> <span class="co"># How many cells per metacell.</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">cluster_id</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span><span class="op">(</span><span class="st">"Computing metacells for cluster %s."</span>, <span class="va">cluster_id</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># Will store the metacells per cluster.</span></span>
<span>  <span class="va">metacells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span>, row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Subset the sample by each cluster ID.</span></span>
<span>  <span class="va">chunksample</span> <span class="op">&lt;-</span> <span class="va">sample</span><span class="op">[</span>, <span class="va">sample</span><span class="op">$</span><span class="va">TME_annotation</span> <span class="op">==</span> <span class="va">cluster_id</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="co"># Get the count data as a data frame and transpose it so columns are GENES and rows are CELLS.</span></span>
<span>  <span class="va">countdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu">GetAssayData</span><span class="op">(</span><span class="va">chunksample</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, slot <span class="op">=</span>  <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Get the possible amount of metacells.</span></span>
<span>  <span class="va">times</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">trunc</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">countdata</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">/</span> <span class="va">metacell_content</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">1</span>,<span class="va">times</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">meta_counter</span> <span class="op">&lt;-</span> <span class="va">meta_counter</span> <span class="op">+</span> <span class="fl">1</span></span>
<span>    <span class="co"># Generate slice points for each metacell.</span></span>
<span>    <span class="va">start</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="op">(</span><span class="va">i</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op">*</span> <span class="va">metacell_content</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="va">end</span> <span class="op">&lt;-</span> <span class="va">i</span> <span class="op">*</span> <span class="va">metacell_content</span></span>
<span>    </span>
<span>    </span>
<span>    <span class="co"># Compute the slice as a data frame containing the sum of the subsetted cells. dims = 1 row (metacell), X columns (genes)</span></span>
<span>    <span class="va">slice</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">countdata</span><span class="op">[</span><span class="va">start</span><span class="op">:</span><span class="va">end</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># Get the name of the cells merged.</span></span>
<span>    <span class="va">cell_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">countdata</span><span class="op">[</span><span class="va">start</span><span class="op">:</span><span class="va">end</span>, <span class="op">]</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># Add the metacell.</span></span>
<span>    <span class="va">col_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span><span class="op">(</span><span class="st">"metacell_%s"</span>, <span class="va">meta_counter</span><span class="op">)</span></span>
<span>    <span class="va">metacells</span><span class="op">[[</span><span class="va">col_name</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">slice</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span>    </span>
<span>    <span class="co"># Add the mapping.</span></span>
<span>    <span class="va">sample</span><span class="op">$</span><span class="va">metacell_mapping</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">cell_names</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">col_name</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="co"># Delete the test column as we already have more than 1 column in our data frame.</span></span>
<span>  <span class="va">metacells</span><span class="op">[[</span><span class="st">"test"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span>  </span>
<span>  <span class="co"># Will contain the annotation of the generated metacells. Columns: cluster identities. Rows: each metacell.</span></span>
<span>  <span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>cluster_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">metacells</span><span class="op">)</span>, row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">metacells</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># Replace the dummy cluster_names column's values for the actual label for the cluster.</span></span>
<span>  <span class="va">annotation</span><span class="op">$</span><span class="va">cluster_names</span> <span class="op">&lt;-</span> <span class="va">cluster_id</span></span>
<span>  </span>
<span>  <span class="co"># Add the annotation data and the metacell data to the global containers. In the end: # Columns for metacell object = # rows for annotation object.</span></span>
<span>  <span class="va">whole_metacells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">whole_metacells</span>, <span class="va">metacells</span><span class="op">)</span></span>
<span>  <span class="va">whole_annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">whole_annotation</span>, <span class="va">annotation</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Turn the names into characters for the sake of avoiding errors when subsetting.</span></span>
<span><span class="va">whole_annotation</span><span class="op">$</span><span class="va">cluster_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">whole_annotation</span><span class="op">$</span><span class="va">cluster_names</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Delete the test row from the global annotation data.</span></span>
<span><span class="va">whole_annotation</span> <span class="op">&lt;-</span> <span class="va">whole_annotation</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">whole_annotation</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"test"</span><span class="op">)</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Delete the test column from the global metacell data.</span></span>
<span><span class="va">whole_metacells</span><span class="op">$</span><span class="va">test</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span></span>
<span><span class="va">cnv_analysis_folder</span> <span class="op">&lt;-</span> <span class="st">""</span> <span class="co"># Path to store the output of inferCNV</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">cnv_analysis_folder</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">annotation_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">cnv_analysis_folder</span>, <span class="st">"/annotation_metacells.tsv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Save the annotation object.</span></span>
<span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">whole_annotation</span>,</span>
<span>                   file <span class="op">=</span> <span class="va">annotation_file</span>,</span>
<span>                   sep <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>                   row.names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                   col.names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                   quote <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Return the metacell object as a matrix (required for running inferCNV).</span></span>
<span><span class="va">count_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">whole_metacells</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run inferCNV:</span></span>
<span><span class="va">gene_ordering_file</span> <span class="op">&lt;-</span> <span class="st">""</span> <span class="co"># Path to where the file with the order of the genes is stored. It can also be downloaded here: https://data.broadinstitute.org/Trinity/CTAT/cnv/</span></span>
<span></span>
<span><span class="va">ref_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="co"># Which clusters to use as a reference.</span></span>
<span></span>
<span><span class="co"># Create the inferCNV object.</span></span>
<span><span class="va">infercnv_obj</span> <span class="op">&lt;-</span> <span class="fu">infercnv</span><span class="fu">::</span><span class="fu">CreateInfercnvObject</span><span class="op">(</span>raw_counts_matrix <span class="op">=</span> <span class="va">count_matrix</span>,</span>
<span>                                               annotations_file <span class="op">=</span> <span class="va">annotation_file</span>,</span>
<span>                                               delim <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>                                               gene_order_file <span class="op">=</span> <span class="va">gene_ordering_file</span>,</span>
<span>                                               ref_group_names <span class="op">=</span> <span class="va">ref_clusters</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run inferCNV.</span></span>
<span><span class="va">cnv_analysis_folder_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">cnv_analysis_folder</span>, <span class="st">"/output"</span><span class="op">)</span> <span class="co"># This path needs to not exist in your filesystem otherwise inferCNV will stop complaining that the path exists.</span></span>
<span></span>
<span><span class="va">infercnv_obj</span> <span class="op">&lt;-</span> <span class="fu">infercnv</span><span class="fu">::</span><span class="fu">run</span><span class="op">(</span><span class="va">infercnv_obj</span>,</span>
<span>                              cutoff <span class="op">=</span> <span class="fl">0.1</span>,  <span class="co"># use 1 for smart-seq, 0.1 for 10x-genomics</span></span>
<span>                              min_cells_per_gene <span class="op">=</span> <span class="fl">3</span>, <span class="co"># Default.</span></span>
<span>                              out_dir <span class="op">=</span> <span class="va">cnv_analysis_folder</span>,  <span class="co"># dir is auto-created for storing outputs</span></span>
<span>                              cluster_by_groups <span class="op">=</span> <span class="cn">TRUE</span>,   <span class="co"># Cluster by groups.</span></span>
<span>                              denoise <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                              HMM <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                              HMM_type <span class="op">=</span> <span class="st">"i6"</span>,</span>
<span>                              window_length <span class="op">=</span> <span class="fl">201</span>,</span>
<span>                              num_threads <span class="op">=</span> <span class="fl">8</span>,</span>
<span>                              resume_mode<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># For further denoising, you can apply median filtering, but this might remove true signal from the plot.</span></span>
<span><span class="va">infercnv_obj_median_filtered</span> <span class="op">=</span> <span class="fu">infercnv</span><span class="fu">::</span><span class="fu">apply_median_filtering</span><span class="op">(</span><span class="va">infercnv_obj</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">infercnv</span><span class="fu">::</span><span class="fu">plot_cnv</span><span class="op">(</span><span class="va">infercnv_obj_median_filtered</span>,</span>
<span>                   out_dir <span class="op">=</span>  <span class="va">cnv_analysis_folder</span>,</span>
<span>                   output_filename <span class="op">=</span> <span class="st">'infercnv.median_filtered'</span>,</span>
<span>                   x.range <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>                   x.center <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                   title <span class="op">=</span> <span class="st">"infercnv"</span>,</span>
<span>                   color_safe_pal <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>



</main><!-- /main --><!--- All credits for this function to: https://stackoverflow.com/a/74089703--><script>
function plot_fold() {
    cells = document.querySelectorAll(".cell:has(img)");
    cells.forEach(cell => {
      imgs = cell.querySelectorAll(".cell-output-display:has(p img)");
      imgs.forEach(img => {
        det = document.createElement("details");
        sum = document.createElement("summary");
        det.appendChild(sum);
        det.children[0].innerText = "Plot";
        img_clone = img.cloneNode(true);
        det.appendChild(img_clone);
        det.open = true;
        cell.replaceChild(det, img);
      });
    });
}

window.onload = plot_fold();
</script><script>
window.onload = function() {
	// Intensify all images on the page.
    var element = document.querySelectorAll( "img");
	  Intense( element );
}
</script><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../../analysis/snRNAseq/03_Normalization_and_Integration.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Normalization and Integration</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../analysis/snRNAseq/05_Supervised_Annotation.html" class="pagination-link">
        <span class="nav-page-text">Supervised Annotation</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">Designed by <strong>Enrique Blanco Carmona</strong></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Figures generated with <strong><a href="https://github.com/enblacar/SCpubr">SCpubr</a></strong> | Book built with <strong><a href="https://quarto.org/">Quarto</a></strong></div>
  </div>
</footer>


<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>