<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.324">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>A cycling, progenitor-like cell population at the base of atypical teratoid rhabdoid tumor subtype differentiation trajectories - snATACseq analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../analysis/Xenium/01_Analysis.html" rel="next">
<link href="../../analysis/snRNAseq/05_Supervised_Annotation.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script><style>html{ scroll-behavior: smooth; }</style>
<script type="text/javascript" src="./intense.js"></script>
<meta property="og:title" content="A cycling, progenitor-like cell population at the base of atypical teratoid rhabdoid tumor subtype differentiation trajectories - snATACseq analysis">
<meta property="og:site-name" content="A cycling, progenitor-like cell population at the base of atypical teratoid rhabdoid tumor subtype differentiation trajectories">
<meta name="twitter:title" content="A cycling, progenitor-like cell population at the base of atypical teratoid rhabdoid tumor subtype differentiation trajectories - snATACseq analysis">
<meta name="twitter:card" content="summary">
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../analysis/snATACseq/01_Analysis.html">Analysis | snATACseq</a></li><li class="breadcrumb-item"><a href="../../analysis/snATACseq/01_Analysis.html">snATACseq analysis</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../../">A cycling, progenitor-like cell population at the base of atypical teratoid rhabdoid tumor subtype differentiation trajectories</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/DrostGroup/ATRT_SC_Het" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false">
 <span class="menu-text">Abstract</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Co-First Authors</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false">
 <span class="menu-text">Analysis | snRNAseq</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/snRNAseq/01_Quality_Control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quality Control</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/snRNAseq/02_Doublet_Removal.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Doublet Removal</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/snRNAseq/03_Normalization_and_Integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Normalization and Integration</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/snRNAseq/04_CNV_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">CNV analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/snRNAseq/05_Supervised_Annotation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Supervised Annotation</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Analysis | snATACseq</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/snATACseq/01_Analysis.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">snATACseq analysis</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">
 <span class="menu-text">Analysis | Xenium</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/Xenium/01_Analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Xenium analysis</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
 <span class="menu-text">Analysis | bulkRNAseq</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/bulkRNAseq/01_GenerateDESeq2Dataset.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Generate DESeq2 Dataset</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/bulkRNAseq/02_QC_Assessment.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quality Control</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../analysis/bulkRNAseq/03_DE_Analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">DE analysis</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false">
 <span class="menu-text">Figures</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure 1</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure 2</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure 3</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure 4</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S1</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S1</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S2</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S4</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S5</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S6.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S6</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S7.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S7</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S8.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S8</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S9.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S9</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S10.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S10</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S11.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S11</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../visualization/Figure_S12.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Figure S12</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">snATACseq analysis</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#This scripts described how to pre-process 10X Genomics multiome data starting with an output of cell-ranger-arc</span></span>
<span></span>
<span><span class="co">#For questions contact the Drost Lab: j.drost@prinsesmaximacentrum.nl</span></span>
<span></span>
<span><span class="co">#Pipeline as described below was used for organoids as well as tissue samples used for scMultiome-seq</span></span>
<span></span>
<span><span class="co">#Libraries needed</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac">Signac</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">JASPAR2020</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TFBSTools</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org">future</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org">Matrix</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org">scales</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://oompa.r-forge.r-project.org/">Polychrome</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">biomaRt</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">GO.db</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://rsqlite.r-dbi.org">RSQLite</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SingleR</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/lazappi/clustree">clustree</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">infercnv</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">celldex</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://had.co.nz/plyr">plyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">chromVAR</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Change to your working directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"/hpc/pmc_drost/rstudio_ire/multiome_IP/libraries"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#plan("multicore", workers = 4)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>future.globals.maxSize <span class="op">=</span> <span class="fl">50</span> <span class="op">*</span> <span class="fl">1024</span> <span class="op">^</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##Load gene sets used to remove confounder effect of cell cycle and red blood cells</span></span>
<span><span class="va">genesRM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"genes_remove_HB_CC_stress.RDS"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Configure variables ----</span></span>
<span><span class="va">basedir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">CFG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">CFG</span><span class="op">$</span><span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">basedir</span>, <span class="st">"/data/"</span><span class="op">)</span></span>
<span><span class="va">CFG</span><span class="op">$</span><span class="va">work_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">basedir</span>, <span class="st">"/work_dir/"</span><span class="op">)</span></span>
<span><span class="va">CFG</span><span class="op">$</span><span class="va">output_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">basedir</span>, <span class="st">"/output/"</span><span class="op">)</span></span>
<span><span class="va">CFG</span><span class="op">$</span><span class="va">metadata_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">basedir</span><span class="op">)</span>       <span class="co">#Needs to be changed depending on where this file is stored</span></span>
<span><span class="va">CFG</span><span class="op">$</span><span class="va">workspace_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">basedir</span>, <span class="st">"/workspace/"</span><span class="op">)</span></span>
<span><span class="va">CFG</span><span class="op">$</span><span class="va">ndims</span> <span class="op">&lt;-</span> <span class="fl">40</span>           <span class="co">#Number of dimensions for dimensional reduction</span></span>
<span><span class="va">CFG</span><span class="op">$</span><span class="va">random_seed</span> <span class="op">&lt;-</span> <span class="fl">1234</span>   <span class="co">#Random seed for UMAP generation</span></span>
<span><span class="va">CFG</span><span class="op">$</span><span class="va">mito_pattern</span> <span class="op">&lt;-</span> <span class="st">"^MT-"</span></span>
<span><span class="va">CFG</span><span class="op">$</span><span class="va">min_txpts</span> <span class="op">&lt;-</span> <span class="fl">800</span>      <span class="co">#As discussed with the single cell facility in the Princess Maxima Center </span></span>
<span><span class="va">CFG</span><span class="op">$</span><span class="va">max_txpts</span> <span class="op">&lt;-</span> <span class="fl">30000</span>    </span>
<span><span class="va">CFG</span><span class="op">$</span><span class="va">min_features</span> <span class="op">&lt;-</span> <span class="fl">500</span>  </span>
<span><span class="va">CFG</span><span class="op">$</span><span class="va">max_pctmito</span> <span class="op">&lt;-</span> <span class="fl">20</span>     </span>
<span><span class="va">CFG</span><span class="op">$</span><span class="va">min_TSS</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">CFG</span><span class="op">$</span><span class="va">max_nucleo</span> <span class="op">&lt;-</span> <span class="fl">1.5</span></span>
<span><span class="va">CFG</span><span class="op">$</span><span class="va">max_frgmts</span> <span class="op">&lt;-</span> <span class="fl">50000</span></span>
<span><span class="va">CFG</span><span class="op">$</span><span class="va">min_frgmts</span> <span class="op">&lt;-</span> <span class="fl">800</span></span>
<span><span class="va">CFG</span><span class="op">$</span><span class="va">max_pct_hemo</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span><span class="va">CFG</span><span class="op">$</span><span class="va">gradient_colors</span> <span class="op">=</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu">viridis</span><span class="op">(</span><span class="fl">101</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">CFG</span><span class="op">$</span><span class="va">cbPalette4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#999999"</span>, <span class="st">"#E69F00"</span>, <span class="st">"#56B4E9"</span>, <span class="st">"#009E73"</span>, <span class="st">"#F0E442"</span>, <span class="st">"#0072B2"</span>, <span class="st">"#EE5050"</span>,</span>
<span>                    <span class="st">"#111111"</span>, <span class="st">"#00E656"</span>, <span class="st">"#C8C8C8"</span>, <span class="st">"#E6009F"</span>, <span class="st">"hotpink1"</span>, <span class="st">"#D55E00"</span>, <span class="st">"#CC79A7"</span>,</span>
<span>                    <span class="st">"cyan2"</span>, <span class="st">"darkmagenta"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="va">CFG</span><span class="op">$</span><span class="va">random_seed</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#Get gene annotations for hg38</span></span>
<span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu">GetGRangesFromEnsDb</span><span class="op">(</span>ensdb <span class="op">=</span> <span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span><span class="va">ucsc.levels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html">str_replace</a></span><span class="op">(</span>string<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"chr"</span>,<span class="fu">seqlevels</span><span class="op">(</span><span class="va">annotation</span><span class="op">)</span>,sep<span class="op">=</span><span class="st">""</span><span class="op">)</span>, pattern<span class="op">=</span><span class="st">"chrMT"</span>, replacement<span class="op">=</span><span class="st">"chrM"</span><span class="op">)</span> <span class="co">#mitochondria annotation</span></span>
<span><span class="fu">seqlevels</span><span class="op">(</span><span class="va">annotation</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">ucsc.levels</span></span>
<span><span class="fu">genome</span><span class="op">(</span><span class="va">annotation</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"hg38"</span></span>
<span></span>
<span></span>
<span><span class="co"># Load data ----</span></span>
<span><span class="co">## Load metadata and create directory information ----</span></span>
<span><span class="va">metadata.exp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'/hpc/pmc_drost/rstudio_ire/multiome_IP/libraries/multiome_drost_metadata_new.txt'</span><span class="op">)</span>, header<span class="op">=</span><span class="cn">T</span><span class="op">)</span> <span class="co">#Loads metadata table for experiments (M) with accompanying library and analysis numbers</span></span>
<span></span>
<span><span class="va">experiments</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">metadata.exp</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span>   <span class="co">#Here you can subset for experiments (M) of interest (or just take all)</span></span>
<span></span>
<span><span class="va">lib</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">metadata.exp</span><span class="op">$</span><span class="va">experiment</span>, <span class="st">"_"</span>,</span>
<span>                     <span class="va">metadata.exp</span><span class="op">$</span><span class="va">RNA_lib</span>, <span class="st">"_"</span>,</span>
<span>                     <span class="va">metadata.exp</span><span class="op">$</span><span class="va">ATAC_lib</span>, <span class="st">"_"</span>,</span>
<span>                     <span class="va">metadata.exp</span><span class="op">$</span><span class="va">analysis</span></span>
<span><span class="op">)</span><span class="op">)</span>  <span class="co">#Generates character vector with the full path name on the surfdrive seq_data folder to identify the libraries per experiment</span></span>
<span></span>
<span><span class="co">#Generates dataframe with the directory for the count install.packages ATAC fragments (column 3) to be retrieved</span></span>
<span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">experiments</span>, </span>
<span>                  dirFilt <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">CFG</span><span class="op">$</span><span class="va">data_dir</span>, <span class="va">lib</span>, <span class="st">"/filtered_feature_bc_matrix"</span><span class="op">)</span>,</span>
<span>                  dirAll <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">CFG</span><span class="op">$</span><span class="va">data_dir</span>, <span class="va">lib</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">select_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">(</span><span class="va">metadata.exp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">source</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"organoid"</span>,<span class="st">"organoids"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="co">#this line kchanges if you take the tissue data</span></span>
<span></span>
<span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="va">dir</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">dir</span><span class="op">$</span><span class="va">experiments</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">select_data</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span></span>
<span><span class="co">## Function to calculate colSums in GEX data per experiment  ---</span></span>
<span><span class="co">## Function to calculate colSums in GEX data per experiment  ---</span></span>
<span><span class="va">fun.GEX.colSums</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">a</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co">#b &lt;- Matrix::colSums(counts.list[[a]]) </span></span>
<span>  <span class="va">b</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html">colSums</a></span><span class="op">(</span><span class="va">counts.list</span><span class="op">[[</span><span class="va">a</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="st">'Gene Expression'</span><span class="op">)</span> <span class="co">##this is for combined analysis</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">b</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># load the RNA and ATAC data and create a Seurat object</span></span>
<span><span class="co">##Here we read only GEX data</span></span>
<span><span class="va">fun.READ10X</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">a</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">counts.list</span> <span class="op">&lt;-</span> <span class="fu">Read10X</span><span class="op">(</span>data.dir <span class="op">=</span> <span class="va">dir</span><span class="op">[</span><span class="va">dir</span><span class="op">$</span><span class="va">experiments</span> <span class="op">==</span> <span class="va">a</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span> </span>
<span>  <span class="co">#counts.list &lt;- counts.list[[1]] ##comment this if running both ATAC and GEX together</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">counts.list</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">fun.GEX.meta</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">a</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">data</span> <span class="op">&lt;-</span> <span class="va">counts.list</span><span class="op">[[</span><span class="va">a</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="st">'Gene Expression'</span>                                <span class="co">#use this if doing combined ATAC and GEX reading</span></span>
<span>  <span class="co">#data &lt;- counts.list[[a]]    </span></span>
<span>  <span class="va">mitos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="va">CFG</span><span class="op">$</span><span class="va">mito_pattern</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>             <span class="co">#Find mitochondrial genes</span></span>
<span>  <span class="va">percent_mito</span> <span class="op">&lt;-</span> <span class="fl">100</span> <span class="op">*</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html">colSums</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span><span class="va">mitos</span>, <span class="op">]</span><span class="op">)</span> <span class="op">/</span> <span class="va">GEX.colSums</span><span class="op">[[</span><span class="va">a</span><span class="op">]</span><span class="op">]</span>   <span class="co">#Calculate percentage of mitochondrial genes per cell;  gives named vector which is necessary for adding the meta.data later (barcodes have to be included)</span></span>
<span>  <span class="va">nuclear</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>, <span class="va">mitos</span><span class="op">)</span>                                 <span class="co">#Define nuclear genes via setdiff function</span></span>
<span>  <span class="va">log2_transcript_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html">colSums</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span><span class="va">nuclear</span> , <span class="op">]</span><span class="op">)</span><span class="op">)</span>     <span class="co">#log2_transcript_counts for nuclear genes</span></span>
<span>  <span class="va">log2_feature_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html">colSums</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span><span class="va">nuclear</span>, <span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>percent_mito <span class="op">=</span> <span class="va">percent_mito</span>,                           <span class="co">#create meta.data df for Seurat object</span></span>
<span>                     log2_counts <span class="op">=</span> <span class="va">log2_transcript_counts</span>,</span>
<span>                     log2_features <span class="op">=</span> <span class="va">log2_feature_counts</span>,</span>
<span>                     source <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="va">metadata.exp</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">a</span>, <span class="va">metadata.exp</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span>,<span class="fl">6</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     lib <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="va">a</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>                              <span class="co">#Automatically adds correct experiment to df</span></span>
<span>  <span class="va">list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">meta</span>, <span class="va">nuclear</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">list</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span></span>
<span><span class="va">fun.GEX.Seurat</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">a</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">counts.list</span><span class="op">[[</span><span class="va">a</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">`Gene Expression`</span><span class="op">[</span><span class="va">GEX.meta</span><span class="op">[[</span><span class="va">a</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span>                               <span class="co">#use this if doing combined ATAC and GEX reading</span></span>
<span>  <span class="co">#counts &lt;- counts.list[[a]]</span></span>
<span>  <span class="va">srat</span> <span class="op">&lt;-</span> <span class="fu">CreateSeuratObject</span><span class="op">(</span>counts <span class="op">=</span> <span class="va">counts</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, meta <span class="op">=</span> <span class="va">GEX.meta</span><span class="op">[[</span><span class="va">a</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>        <span class="co">#Add information in meta.data from above</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">srat</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">## Load data for selected data/library ----</span></span>
<span><span class="va">counts.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">select_data</span>, <span class="va">fun.READ10X</span><span class="op">)</span>     <span class="co">#Make sure to use the correct function (both, GEX, ATAC)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">counts.list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">select_data</span></span>
<span><span class="va">GEX.colSums</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">select_data</span>, <span class="va">fun.GEX.colSums</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">GEX.colSums</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">select_data</span></span>
<span></span>
<span><span class="co">## Generate GEX meta.data list ----</span></span>
<span><span class="va">GEX.meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">select_data</span>, <span class="va">fun.GEX.meta</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">GEX.meta</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">select_data</span></span>
<span><span class="co">##Create Seurat object with data and metadata</span></span>
<span><span class="va">srat.GEX</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">select_data</span>, <span class="va">fun.GEX.Seurat</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create  ATAC assay and add it to the object</span></span>
<span><span class="co">##But first, read peak files</span></span>
<span><span class="va">grFiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">exp</span> <span class="kw">in</span> <span class="va">select_data</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># ##read peaks file</span></span>
<span>  <span class="va">bedFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">dir</span><span class="op">$</span><span class="va">dirAll</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">exp</span>, <span class="va">dir</span><span class="op">$</span><span class="va">experiments</span><span class="op">)</span><span class="op">]</span>, <span class="st">"/atac_peaks.bed"</span><span class="op">)</span>,</span>
<span>                        col.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="st">"start"</span>, <span class="st">"end"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span>  <span class="va">grFiles</span><span class="op">[[</span><span class="va">exp</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">makeGRangesFromDataFrame</span><span class="op">(</span><span class="va">bedFile</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># Create a unified set of peaks to quantify in each dataset</span></span>
<span><span class="va">combined.peaks</span> <span class="op">&lt;-</span> <span class="fu">disjoin</span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">grFiles</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">grFiles</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="va">grFiles</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Filter out bad peaks based on length and combine peaks</span></span>
<span><span class="va">peakwidths</span> <span class="op">&lt;-</span> <span class="fu">width</span><span class="op">(</span><span class="va">combined.peaks</span><span class="op">)</span></span>
<span><span class="va">combined.peaks</span> <span class="op">&lt;-</span> <span class="va">combined.peaks</span><span class="op">[</span><span class="va">peakwidths</span>  <span class="op">&lt;</span> <span class="fl">10000</span> <span class="op">&amp;</span> <span class="va">peakwidths</span> <span class="op">&gt;</span> <span class="fl">20</span><span class="op">]</span></span>
<span><span class="co">##Here filter further to only keep standard chromosomes</span></span>
<span><span class="va">combined.peaks</span> <span class="op">&lt;-</span> <span class="fu">keepStandardChromosomes</span><span class="op">(</span><span class="va">combined.peaks</span>, pruning.mode<span class="op">=</span><span class="st">"coarse"</span><span class="op">)</span></span>
<span><span class="co">##Here filter to remove blacklisted genes</span></span>
<span><span class="va">combined.peaks</span> <span class="op">&lt;-</span> <span class="fu">subsetByOverlaps</span><span class="op">(</span>x <span class="op">=</span> <span class="va">combined.peaks</span>, ranges <span class="op">=</span> <span class="va">blacklist_hg38_unified</span>, invert <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">atacAll</span> <span class="op">&lt;-</span> <span class="va">srat.GEX</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">exp</span> <span class="kw">in</span> <span class="va">select_data</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">fragpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">dir</span><span class="op">$</span><span class="va">dirAll</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">exp</span>, <span class="va">dir</span><span class="op">$</span><span class="va">experiments</span><span class="op">)</span><span class="op">]</span>, <span class="st">"/atac_fragments.tsv.gz"</span><span class="op">)</span></span>
<span>  <span class="va">atacAll</span><span class="op">[[</span><span class="va">exp</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"ATAC"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">CreateChromatinAssay</span><span class="op">(</span></span>
<span>    counts <span class="op">=</span> <span class="va">counts.list</span><span class="op">[[</span><span class="va">exp</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Peaks</span>,</span>
<span>    sep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">":"</span>, <span class="st">"-"</span><span class="op">)</span>,</span>
<span>    fragments <span class="op">=</span> <span class="va">fragpath</span>,</span>
<span>    annotation <span class="op">=</span> <span class="va">annotation</span></span>
<span>  <span class="op">)</span></span>
<span>  <span class="fu">DefaultAssay</span><span class="op">(</span><span class="va">atacAll</span><span class="op">[[</span><span class="va">exp</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"ATAC"</span></span>
<span>  <span class="va">atacAll</span><span class="op">[[</span><span class="va">exp</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">NucleosomeSignal</span><span class="op">(</span><span class="va">atacAll</span><span class="op">[[</span><span class="va">exp</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">atacAll</span><span class="op">[[</span><span class="va">exp</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">TSSEnrichment</span><span class="op">(</span><span class="va">atacAll</span><span class="op">[[</span><span class="va">exp</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="co">##Create fragment object for each sample</span></span>
<span>  <span class="va">frags</span> <span class="op">&lt;-</span> <span class="fu">CreateFragmentObject</span><span class="op">(</span>path <span class="op">=</span> <span class="va">fragpath</span><span class="op">)</span></span>
<span>  <span class="va">CRcounts</span> <span class="op">&lt;-</span> <span class="fu">FeatureMatrix</span><span class="op">(</span>fragments <span class="op">=</span><span class="fu">Fragments</span><span class="op">(</span><span class="va">atacAll</span><span class="op">[[</span><span class="va">exp</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                            features <span class="op">=</span> <span class="va">combined.peaks</span>, </span>
<span>                            cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">atacAll</span><span class="op">[[</span><span class="va">exp</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">atacAll</span><span class="op">[[</span><span class="va">exp</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"peaks"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">CreateChromatinAssay</span><span class="op">(</span></span>
<span>    counts <span class="op">=</span> <span class="va">CRcounts</span>,</span>
<span>    fragments <span class="op">=</span> <span class="va">fragpath</span>,</span>
<span>    annotation <span class="op">=</span> <span class="va">annotation</span></span>
<span>  <span class="op">)</span></span>
<span>  </span>
<span>  </span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html">pdf</a></span><span class="op">(</span>file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">CFG</span><span class="op">$</span><span class="va">output_dir</span>, <span class="va">exp</span>, <span class="st">"_QC_Seurat_merge_blacklist.pdf"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">VlnPlot</span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">atacAll</span><span class="op">[[</span><span class="va">exp</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nCount_RNA"</span>, <span class="st">"nCount_ATAC"</span>, <span class="st">"TSS.enrichment"</span>, <span class="st">"nucleosome_signal"</span><span class="op">)</span>,</span>
<span>    ncol <span class="op">=</span> <span class="fl">4</span>,</span>
<span>    pt.size <span class="op">=</span> <span class="fl">0</span></span>
<span>  <span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">atacAll</span>, <span class="st">"srat_merged_disjoin_organoids.RDS"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">exp</span> <span class="kw">in</span> <span class="va">select_data</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">atacAll</span><span class="op">[[</span><span class="va">exp</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">atacAll</span><span class="op">[[</span><span class="va">exp</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    subset <span class="op">=</span> <span class="va">nCount_RNA</span> <span class="op">&gt;=</span> <span class="va">CFG</span><span class="op">$</span><span class="va">min_txpts</span> <span class="op">&amp;</span>    <span class="co">#Minimum threshold transcripts </span></span>
<span>      <span class="va">nCount_RNA</span> <span class="op">&lt;=</span> <span class="va">CFG</span><span class="op">$</span><span class="va">max_txpts</span> <span class="op">&amp;</span>     <span class="co">#Maximum threshold transcripts</span></span>
<span>      <span class="va">percent_mito</span> <span class="op">&lt;=</span> <span class="va">CFG</span><span class="op">$</span><span class="va">max_pctmito</span> <span class="op">&amp;</span></span>
<span>      <span class="va">nucleosome_signal</span> <span class="op">&lt;</span> <span class="va">CFG</span><span class="op">$</span><span class="va">max_nucleo</span> <span class="op">&amp;</span></span>
<span>      <span class="va">TSS.enrichment</span> <span class="op">&gt;</span> <span class="va">CFG</span><span class="op">$</span><span class="va">min_TSS</span> <span class="op">&amp;</span></span>
<span>      <span class="va">nCount_ATAC</span> <span class="op">&lt;</span> <span class="va">CFG</span><span class="op">$</span><span class="va">max_frgmts</span> <span class="op">&amp;</span></span>
<span>      <span class="va">nCount_ATAC</span> <span class="op">&gt;</span> <span class="va">CFG</span><span class="op">$</span><span class="va">min_frgmts</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html">pdf</a></span><span class="op">(</span>file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">CFG</span><span class="op">$</span><span class="va">output_dir</span>, <span class="va">exp</span>, <span class="st">"_QC_filtered_merged_BL.pdf"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">VlnPlot</span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">atacAll</span><span class="op">[[</span><span class="va">exp</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nCount_RNA"</span>, <span class="st">"nCount_ATAC"</span>, <span class="st">"TSS.enrichment"</span>, <span class="st">"nucleosome_signal"</span><span class="op">)</span>,</span>
<span>    ncol <span class="op">=</span> <span class="fl">4</span>,</span>
<span>    pt.size <span class="op">=</span> <span class="fl">0</span></span>
<span>  <span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span></span>
<span><span class="co">##Merge seurat objects from the filtered libraries</span></span>
<span><span class="va">srat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span><span class="va">atacAll</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">atacAll</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="va">atacAll</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>              add.cell.ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">atacAll</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Normalize GEX data using LogNormalize ----</span></span>
<span><span class="fu">DefaultAssay</span><span class="op">(</span><span class="va">srat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"RNA"</span></span>
<span><span class="va">srat</span> <span class="op">&lt;-</span> <span class="fu">NormalizeData</span><span class="op">(</span><span class="va">srat</span>, normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span><span class="op">)</span>    <span class="co">#Goes to slot "Data" of RNA assay (GetAssayData function)</span></span>
<span><span class="va">srat</span> <span class="op">&lt;-</span> <span class="fu">ScaleData</span><span class="op">(</span><span class="va">srat</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">srat</span><span class="op">)</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>   <span class="co">#Goes to slot "scale.data" of RNA assay (GetAssayData function)</span></span>
<span><span class="va">srat</span> <span class="op">&lt;-</span> <span class="fu">FindVariableFeatures</span><span class="op">(</span><span class="va">srat</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Normalize GEX data using SCTransform ----</span></span>
<span><span class="co"># SCTransform automatically scales data and finds variables</span></span>
<span><span class="fu">DefaultAssay</span><span class="op">(</span><span class="va">srat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"RNA"</span></span>
<span><span class="va">srat</span> <span class="op">&lt;-</span> <span class="fu">SCTransform</span><span class="op">(</span><span class="va">srat</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span>, variable.features.n <span class="op">=</span> <span class="fl">3000</span><span class="op">)</span>  <span class="co">#Specify number of features for downstream analysis</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">srat</span>, <span class="st">"090523_srat_SCT_done"</span><span class="op">)</span></span>
<span><span class="va">srat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"120523_srat_SCT_done_organoids"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Inspect most highly variable genes ----</span></span>
<span><span class="fu">DefaultAssay</span><span class="op">(</span><span class="va">srat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"SCT"</span></span>
<span><span class="va">top10</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">VariableFeatures</span><span class="op">(</span><span class="va">srat</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu">VariableFeaturePlot</span><span class="op">(</span><span class="va">srat</span><span class="op">)</span></span>
<span><span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu">LabelPoints</span><span class="op">(</span>plot <span class="op">=</span> <span class="va">plot1</span>, points <span class="op">=</span> <span class="va">top10</span><span class="op">)</span></span>
<span><span class="va">plot2</span></span>
<span></span>
<span><span class="co">##Dimensional reduction ----</span></span>
<span><span class="co">### PCA: Calculation ----</span></span>
<span><span class="co"># There can only be one PCA in the Seurat object. Start with SCTransform to generate UMAP; LogNormalized will be used for gene expression analysis</span></span>
<span><span class="va">srat</span> <span class="op">&lt;-</span> <span class="fu">RunPCA</span><span class="op">(</span><span class="va">srat</span>, npcs <span class="op">=</span><span class="fl">50</span><span class="op">)</span>                             <span class="co">#PCA for SCTransformed Seurat object</span></span>
<span><span class="co">### PCA: Visualize nuclei along the first two principal axes per normalization method ----</span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">srat</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, pt.size <span class="op">=</span> <span class="fl">1</span>, group.by <span class="op">=</span> <span class="st">'lib'</span>, cols <span class="op">=</span> <span class="va">CFG</span><span class="op">$</span><span class="va">cbPalette4</span><span class="op">)</span></span>
<span><span class="co">### Elbow Plot to choose number of dimensions ----</span></span>
<span><span class="fu">ElbowPlot</span><span class="op">(</span><span class="va">srat</span>, ndims <span class="op">=</span> <span class="fl">50</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span>
<span><span class="co">###Alternatively, consider a more quantitative approach to determine how many pcs to use</span></span>
<span><span class="co"># Determine percent of variation associated with each PC</span></span>
<span><span class="va">pct</span> <span class="op">&lt;-</span> <span class="va">srat</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">pca</span><span class="op">@</span><span class="va">stdev</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">srat</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">pca</span><span class="op">@</span><span class="va">stdev</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span>
<span><span class="co"># Calculate cumulative percents for each PC</span></span>
<span><span class="va">cum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html">cumsum</a></span><span class="op">(</span><span class="va">pct</span><span class="op">)</span></span>
<span><span class="co"># Determine which PC exhibits cumulative percent greater than 90% and % variation associated with the PC as less than 5</span></span>
<span><span class="va">co1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">cum</span> <span class="op">&gt;</span> <span class="fl">90</span> <span class="op">&amp;</span> <span class="va">pct</span> <span class="op">&lt;</span> <span class="fl">5</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="co"># Determine the difference between variation of PC and subsequent PC</span></span>
<span><span class="va">co2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="op">(</span><span class="va">pct</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">pct</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">-</span> <span class="va">pct</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">pct</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.1</span><span class="op">)</span>,  decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">+</span> <span class="fl">1</span> <span class="co"># last point where change of % of variation is more than 0.1%.</span></span>
<span><span class="co"># Minimum of the two calculation</span></span>
<span><span class="va">pcs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">co1</span>, <span class="va">co2</span><span class="op">)</span> <span class="co"># change to any other number</span></span>
<span></span>
<span><span class="co">### PC scores heatmaps ----</span></span>
<span><span class="fu">DimHeatmap</span><span class="op">(</span><span class="va">srat</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, cells <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### UMAP (before filtering) ----</span></span>
<span><span class="va">srat</span> <span class="op">&lt;-</span> <span class="fu">RunUMAP</span><span class="op">(</span><span class="va">srat</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, n.neighbors <span class="op">=</span> <span class="fl">30</span><span class="op">)</span>          <span class="co">#Default n.neighbors = 30</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html">pdf</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"Unfiltered variable features UMAP_merged.pdf"</span><span class="op">)</span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">srat</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span>, group.by <span class="op">=</span> <span class="st">'lib'</span>, cols <span class="op">=</span> <span class="va">CFG</span><span class="op">$</span><span class="va">cbPalette4</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"SCT unfiltered UMAP, 20 PCs"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## Filter certain gene groups from variable features ----</span></span>
<span><span class="co"># This concerns cell cycle genes, stress genes etc. </span></span>
<span><span class="co">##It makes use of the SCutils package developed by our scGenomics facility</span></span>
<span><span class="co">### Remove cells with high hemoglobin content ----</span></span>
<span><span class="va">hb.genes</span> <span class="op">&lt;-</span> <span class="va">genesRM</span><span class="op">$</span><span class="va">hb.genes</span></span>
<span><span class="va">hb.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html">intersect</a></span><span class="op">(</span><span class="va">hb.genes</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">srat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">hb.counts</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html">colSums</a></span><span class="op">(</span><span class="va">srat</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span><span class="op">[</span><span class="va">hb.genes</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="va">srat</span> <span class="op">&lt;-</span> <span class="fu">AddMetaData</span><span class="op">(</span><span class="va">srat</span>,</span>
<span>                    col.name <span class="op">=</span> <span class="st">"log2hb_genes"</span>,</span>
<span>                    metadata <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="va">hb.counts</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">srat</span> <span class="op">&lt;-</span> <span class="fu">AddMetaData</span><span class="op">(</span><span class="va">srat</span>,</span>
<span>                    col.name <span class="op">=</span> <span class="st">"pct_hemo"</span>,</span>
<span>                    metadata <span class="op">=</span> <span class="fl">100</span> <span class="op">*</span> <span class="va">hb.counts</span><span class="op">/</span><span class="va">srat</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">nCount_RNA</span><span class="op">)</span></span>
<span><span class="va">srat</span> <span class="op">&lt;-</span> <span class="fu">AddMetaData</span><span class="op">(</span><span class="va">srat</span>,</span>
<span>                    col.name <span class="op">=</span> <span class="st">"percent.rb"</span>,</span>
<span>                    metadata <span class="op">=</span> <span class="fu">PercentageFeatureSet</span><span class="op">(</span><span class="va">srat</span>, pattern <span class="op">=</span> <span class="st">"^RP[SL]"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">srat</span> <span class="op">&lt;-</span> <span class="fu">CellCycleScoring</span><span class="op">(</span>object <span class="op">=</span> <span class="va">srat</span>,</span>
<span>                         g2m.features <span class="op">=</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">g2m.genes</span>,</span>
<span>                         s.features <span class="op">=</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">s.genes</span><span class="op">)</span></span>
<span></span>
<span><span class="va">srat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">srat</span>, <span class="va">pct_hemo</span> <span class="op">&lt;=</span> <span class="va">CFG</span><span class="op">$</span><span class="va">max_pct_hemo</span><span class="op">)</span>  <span class="co">#Will likely only filter out few nuclei</span></span>
<span></span>
<span><span class="co">#Now we have a filtered dataset where we need to add a chromvar assay </span></span>
<span><span class="co">#Add a chromVAR array to the seurat object </span></span>
<span><span class="co"># Get a list of motif position frequency matrices from the JASPAR database</span></span>
<span><span class="va">pfm</span> <span class="op">&lt;-</span> <span class="fu">getMatrixSet</span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">JASPAR2020</span>,</span>
<span>  opts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>collection <span class="op">=</span> <span class="st">"CORE"</span>,species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, tax_group <span class="op">=</span> <span class="st">'vertebrates'</span>, all_versions <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add motif information</span></span>
<span><span class="fu">DefaultAssay</span><span class="op">(</span><span class="va">srat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'peaks'</span></span>
<span><span class="va">tissue</span> <span class="op">&lt;-</span> <span class="fu">SetIdent</span><span class="op">(</span><span class="va">srat</span>, value<span class="op">=</span><span class="st">"cell_status_and_subtype"</span><span class="op">)</span></span>
<span><span class="va">tissue</span> <span class="op">&lt;-</span> <span class="fu">AddMotifs</span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">tissue</span>,</span>
<span>  genome <span class="op">=</span> <span class="va">BSgenome.Hsapiens.UCSC.hg38</span>,</span>
<span>  pfm <span class="op">=</span> <span class="va">pfm</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#Add chromVar assay to the seurat object</span></span>
<span><span class="fu">register</span><span class="op">(</span><span class="fu">MulticoreParam</span><span class="op">(</span><span class="fl">8</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">register</span><span class="op">(</span><span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">DefaultAssay</span><span class="op">(</span><span class="va">srat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"peaks"</span></span>
<span></span>
<span><span class="co"># Calculate chromvar scores on the peak assay</span></span>
<span><span class="co">#Run this as seperate script as job instead of in the command line </span></span>
<span><span class="va">tissue</span> <span class="op">&lt;-</span> <span class="fu">RunChromVAR</span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">srat</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">'peaks'</span>,</span>
<span>  genome <span class="op">=</span> <span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>



</main><!-- /main --><!--- All credits for this function to: https://stackoverflow.com/a/74089703--><script>
function plot_fold() {
    cells = document.querySelectorAll(".cell:has(img)");
    cells.forEach(cell => {
      imgs = cell.querySelectorAll(".cell-output-display:has(p img)");
      imgs.forEach(img => {
        det = document.createElement("details");
        sum = document.createElement("summary");
        det.appendChild(sum);
        det.children[0].innerText = "Plot";
        img_clone = img.cloneNode(true);
        det.appendChild(img_clone);
        det.open = true;
        cell.replaceChild(det, img);
      });
    });
}

window.onload = plot_fold();
</script><script>
window.onload = function() {
	// Intensify all images on the page.
    var element = document.querySelectorAll( "img");
	  Intense( element );
}
</script><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../../analysis/snRNAseq/05_Supervised_Annotation.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Supervised Annotation</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../analysis/Xenium/01_Analysis.html" class="pagination-link">
        <span class="nav-page-text">Xenium analysis</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">Designed by <strong>Enrique Blanco Carmona</strong></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Figures generated with <strong><a href="https://github.com/enblacar/SCpubr">SCpubr</a></strong> | Book built with <strong><a href="https://quarto.org/">Quarto</a></strong></div>
  </div>
</footer>


<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>